### PR TITLE
[RF] Implement binned fit optimization in BatchMode and other optimizations

### PR DESCRIPTION
**Pull request by guitargeek**
Before, this was done by recursing throught the computation graph each
time a node was evaluated, which was way too expensive for large models.

### CODE REVIEW

File: roofit/roofitcore/inc/RooFit/Detail/DataMap.h [hunk 1]
```diff
@@ -56,6 +58,21 @@ class DataKey {
...
 };
 
+} // namespace Detail
+} // namespace RooFit
+
+namespace std {
+
+template <>
+struct hash<RooFit::Detail::DataKey> {
+   std::size_t operator()(const RooFit::Detail::DataKey &k) const { return hash<TObject const *>()(&*k); }
```
Comment (hageboeck) [inline roofit/roofitcore/inc/RooFit/Detail/DataMap.h:70]: Not `return hash<...>(&k);`

Comment (guitargeek) [reply roofit/roofitcore/inc/RooFit/Detail/DataMap.h:70]: Right, that's more concise! I'll update

**Comment by guitargeek** [comment]
@phsft-bot build

**Review by lmoneta — APPROVED** [review]
Thank you Jonas for implementing this big improvement, LGTM !

**Pull request merged**